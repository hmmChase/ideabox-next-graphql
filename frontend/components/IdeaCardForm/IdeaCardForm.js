import { gql } from 'apollo-boost';
import styled from 'styled-components';
import { Mutation } from 'react-apollo';
import { ALL_IDEAS_QUERY } from '../IdeaContainer/IdeaContainer';

export const CREATE_IDEA_MUTATION = gql`
  mutation CREATE_IDEA_MUTATION($idea: String!) {
    createIdea(idea: $idea) {
      id
    }
  }
`;

class IdeaCardForm extends React.Component {
  state = {
    idea: '',
    isSubmitDisabled: true
  };

  canSubmit = () => {
    if (this.state.idea === '') {
      this.setState({ isSubmitDisabled: true });
    } else {
      this.setState({ isSubmitDisabled: false });
    }
  };

  handleChangeIdeaInput = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value }, () => this.canSubmit());
  };

  handleSubmitIdeaForm = async (event, createIdea) => {
    event.preventDefault();
    this.setState({ isSubmitDisabled: true });
    createIdea();
    this.setState({ idea: '' }, () => this.canSubmit());
  };

  render() {
    return (
      <Mutation
        mutation={CREATE_IDEA_MUTATION}
        variables={{ idea: this.state.idea }}
        refetchQueries={[{ query: ALL_IDEAS_QUERY }]}
      >
        {createIdea => (
          <StyledForm onSubmit={event => this.handleSubmitIdeaForm(event, createIdea)}>
            <img src="static/ideabox.png" alt="ideabox" />
            <textarea
              name="idea"
              type="text"
              placeholder="What's on your mind?"
              value={this.state.idea}
              onChange={event => this.handleChangeIdeaInput(event)}
            />
            <button type="submit" disabled={this.state.isSubmitDisabled}>
              Add Idea
            </button>
          </StyledForm>
        )}
      </Mutation>
    );
  }
}

export default IdeaCardForm;

const StyledForm = styled.form`
  display: flex;
  width: 80%;
  max-width: 800px;
  position: relative;
  img {
    width: 100px;
    position: absolute;
    top: -60px;
    left: -45px;
  }
  textarea {
    padding: 10px;
    outline: none;
    width: 100%;
    padding-left: 20px;
    resize: vertical;
  }
  button {
    width: 60px;
    border: none;
    background-color: #fffacd;
  }
`;
