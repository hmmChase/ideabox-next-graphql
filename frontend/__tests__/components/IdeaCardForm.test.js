import React from 'react';
import { shallow } from 'enzyme';
import IdeaCardForm from '../../components/IdeaCardForm/IdeaCardForm';

describe('IdeaCardForm', () => {
  let mockProps;
  let ideaCardForm;

  beforeEach(() => {
    mockProps = {};
    ideaCardForm = shallow(<IdeaCardForm {...mockProps} />);
  });

  it('matches snapshot', () => {
    expect(ideaCardForm).toMatchSnapshot();
  });
});
