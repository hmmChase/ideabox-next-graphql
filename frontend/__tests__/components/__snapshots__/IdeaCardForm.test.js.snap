// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdeaCardForm matches snapshot 1`] = `ShallowWrapper {}`;
